clab/general.go

```diff
@@ -0,0 +1,126 @@
+package clab
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/cloudflare/cfssl/log"
+	"github.com/srl-labs/containerlab/types"
+)
+
+const (
+	clabHostEntryPrefix  = "###### CLAB-%s-START ######"
+	clabHostEntryPostfix = "###### CLAB-%s-END ######"
+)
+
+func AppendHostsFileEntries(containers []types.GenericContainer, labname string) error {
+	if labname == "" {
+		return fmt.Errorf("missing lab name")
+	}
+	// lets make sure we do not have remaining of a non destroyed run in the hosts file
+	err := DeleteEntriesFromHostsFile(labname)
+	if err != nil {
+		return err
+	}
+	data := generateHostsEntries(containers, labname)
+	if len(data) == 0 {
+		return nil
+	}
+	f, err := os.OpenFile("/etc/hosts", os.O_APPEND|os.O_WRONLY, os.ModeAppend)
```

Comment:
Create if not existing would look like this.

```suggestion
	f, err := os.OpenFile("/etc/hosts", os.O_APPEND|os.O_WRONLY, os.ModeAppend)
	
		if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			f, err = os.Create("/etc/hosts")
			if err != nil {
				return err
			}
			err = f.Chmod(0644)
			if err != nil {
				return err
			}
			err = f.Chown(0, 0)
			if err != nil {
				return err
			}
		} else {
			return err
		}
	}
```

Reply:
If AppendHostEntries appends or creates the file., you can just return here and let AppenHostEntries do it's job. No need to create it here. Just retrun `nil` for AppendHostEntries

Reply:
a basic "|os.O_CREATE" created the file with a permission set of 000. So the perms should be changed. I only get to know when the perms should be changed if I get the error. Hence I need to separately create the file and set the perms.

Reply:
Let's do the following

1. in AppendHosts we check if the `/etc/hosts` is there. We can use already available func https://github.com/srl-labs/containerlab/blob/7e8ea72915b6566ede6cd93f737cefdcd239c5d8/utils/file.go#L14
2. then if it is not there, we create a file (again with the existing func https://github.com/srl-labs/containerlab/blob/7e8ea72915b6566ede6cd93f737cefdcd239c5d8/utils/file.go#L78), with content `127.0.0.1 localhost`
3. with that, we do not mess too much with permissions and reuse the existing code

thoughts, @kellerza @steiler

Reply:
That looks alright to me.

```
root@centos-srlinux:/home/ubuntu/containerlab# go build ; rm /etc/hosts; ls -la /etc/hosts ; ./containerlab  deploy -t tests/01-smoke/01-linux-nodes.clab.yml ; cat /etc/hosts ; ./containerlab  destroy -t tests/01-smoke/01-linux-nodes.clab.yml  ; ls -la /etc/hosts ;cat /etc/hosts
ls: cannot access '/etc/hosts': No such file or directory
INFO[0000] Parsing & checking topology file: 01-linux-nodes.clab.yml
INFO[0000] Creating lab directory: /home/ubuntu/containerlab/clab-2-linux-nodes
INFO[0000] Creating container: l1
INFO[0000] Creating container: l3
INFO[0000] Creating container: l2
INFO[0001] Creating virtual wire: l1:eth1 <--> l2:eth1
INFO[0001] Creating virtual wire: l1:eth2 <--> l2:eth2
INFO[0001] Writing /etc/hosts file
INFO[0001] � New containerlab version 0.15.3 is available! Release notes: https://containerlab.srlinux.dev/rn/0.15/#0.15.3
Run 'containerlab version upgrade' to upgrade or go check other installation options at https://containerlab.srlinux.dev/install/
+---+-----------------------+--------------+---------------------+-------+-------+---------+------------------+------------------------+
| # |         Name          | Container ID |        Image        | Kind  | Group |  State  |   IPv4 Address   |      IPv6 Address      |
+---+-----------------------+--------------+---------------------+-------+-------+---------+------------------+------------------------+
| 1 | clab-2-linux-nodes-l1 | 532fb61a9a53 | alpine:3            | linux |       | running | 172.20.20.2/24   | 2001:172:20:20::2/64   |
| 2 | clab-2-linux-nodes-l2 | 574f0cf19f2a | nginx:stable-alpine | linux |       | running | 172.20.20.100/24 | 2001:172:20:20::100/64 |
| 3 | clab-2-linux-nodes-l3 | 19cbcea3faa7 | nginx:stable-alpine | linux |       | running |                  |                        |
+---+-----------------------+--------------+---------------------+-------+-------+---------+------------------+------------------------+
127.0.0.1       localhost
###### CLAB-2-linux-nodes-START ######
172.20.20.100   clab-2-linux-nodes-l2
172.20.20.2     clab-2-linux-nodes-l1
2001:172:20:20::100     clab-2-linux-nodes-l2
2001:172:20:20::2       clab-2-linux-nodes-l1
###### CLAB-2-linux-nodes-END ######
INFO[0000] Parsing & checking topology file: 01-linux-nodes.clab.yml
INFO[0000] Destroying lab: 2-linux-nodes
INFO[0000] Removed container: clab-2-linux-nodes-l3
INFO[0000] Removed container: clab-2-linux-nodes-l1
INFO[0000] Removed container: clab-2-linux-nodes-l2
INFO[0000] Removing container entries from /etc/hosts file
-rw-r--r-- 1 root root 20 Jul 20 14:36 /etc/hosts
127.0.0.1       localhost
```